inferno-ml-server-types/src/Inferno/ML/Server/Types.hs

```diff
@@ -0,0 +1,762 @@
+import Data.Aeson.Types (Parser)
+import qualified Data.Attoparsec.ByteString.Char8 as Attoparsec
+import Data.Bool (bool)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Char8 as ByteString.Char8
+import Data.Data (Typeable)
+import Data.Generics.Product (HasType (typed), the)
+import Data.Generics.Wrapped (wrappedTo)
+import qualified Data.IP
+import Data.Int (Int64)
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.List.NonEmpty as NonEmpty
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Ord (comparing)
+import Data.Scientific (toRealFloat)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text.Encoding
+import Data.Vector (Vector)
+import qualified Data.Vector as Vector
+import Data.Word (Word32, Word64)
+import Database.PostgreSQL.Simple.FromField
+  ( Conversion,
+    Field,
+    FromField (fromField),
+    ResultError (ConversionFailed, UnexpectedNull),
+    returnError,
+  )
+import Database.PostgreSQL.Simple.FromRow (FromRow (fromRow), field)
+import Database.PostgreSQL.Simple.LargeObjects (Oid (Oid))
+import Database.PostgreSQL.Simple.Newtypes (Aeson (Aeson), getAeson)
+import Database.PostgreSQL.Simple.ToField
+  ( Action (Escape, EscapeByteA),
+    ToField (toField),
+  )
+import Database.PostgreSQL.Simple.ToRow (ToRow (toRow))
+import Database.PostgreSQL.Simple.Types
+  ( Binary (Binary),
+    Default (Default),
+  )
+import Foreign.C (CUInt (CUInt))
+import GHC.Generics (Generic)
+import Inferno.Types.VersionControl
+  ( VCObjectHash,
+    byteStringToVCObjectHash,
+    vcObjectHashToByteString,
+  )
+import Inferno.VersionControl.Types (VCMeta, VCObject)
+import Lens.Micro.Platform hiding ((.=))
+import Servant
+  ( Capture,
+    Get,
+    JSON,
+    NewlineFraming,
+    Post,
+    Put,
+    QueryParam,
+    QueryParam',
+    ReqBody,
+    Required,
+    StreamBody,
+    (:<|>),
+    (:>),
+  )
+import Servant.Conduit ()
+import System.Posix (EpochTime)
+import Text.Read (readMaybe)
+import URI.ByteString (Absolute, URIRef)
+import URI.ByteString.Aeson ()
+import Web.HttpApiData (FromHttpApiData (parseUrlPiece), ToHttpApiData (toUrlPiece))
+
+-- API type for `inferno-ml-server`
+type InfernoMlServerAPI uid gid p s =
+  -- Check if the server is up and if any job is currently running:
+  --
+  --  * `Nothing` -> The server is evaluating a script
+  --  * `Just ()` -> The server is not doing anything and can be killed
+  --
+  -- This can be implemented using an `MVar ()`
+  "status" :> Get '[JSON] (Maybe ())
+    -- Evaluate an inference script
+    :<|> "inference"
+      :> Capture "id" (Id (InferenceParam uid gid p s))
+      :> QueryParam "res" Int64
+      :> Post '[JSON] ()
+    :<|> "inference" :> "cancel" :> Put '[JSON] ()
+    -- Register the bridge. This is an `inferno-ml-server` endpoint, not a
+    -- bridge endpoint
+    :<|> "bridge" :> ReqBody '[JSON] BridgeInfo :> Post '[JSON] ()
```

Comment:
We might want to just embed the bridge info directly into the servers (if we use a stable domain name, for example). But doing it this way allows for the bridge info to be updated dynamically (i.e. while the `inferno-ml-server` is running)